### PR TITLE
Well libs

### PR DESCRIPTION
**Pull request by mgcam**

### CODE REVIEW

File: lang_qc/db/mlwh_schema.py [hunk 1]
```diff
@@ -25,9 +25,30 @@
...
+from sqlalchemy.orm import DeclarativeBase, relationship
 
-Base = declarative_base()
+
+class Base(DeclarativeBase):
+    """
+    A base class for declarative class definitions for the ml warehouse database.
+    """
+
+    def get_row_description(self, fields: list[str]) -> str:
```
Comment (nerdstrike) [inline lang_qc/db/mlwh_schema.py:36]: Maybe make this guy private?

File: lang_qc/models/pacbio/experiment.py [hunk 2]
```diff
@@ -19,14 +19,96 @@
...
-from pydantic import BaseModel, ConfigDict, Field
+from pydantic import Field, model_validator
+from pydantic.dataclasses import dataclass
 
 from lang_qc.db.mlwh_schema import PacBioRun
 
 
-class PacBioExperiment(BaseModel):
+@dataclass(kw_only=True, frozen=True)
+class PacBioLibrary:
```
Comment (nerdstrike) [inline lang_qc/models/pacbio/experiment.py:31]: This isn't a million miles away from being just `Library`, subject to warehouse table resolution.

File: lang_qc/models/pacbio/experiment.py [hunk 3]
```diff
@@ -19,14 +19,96 @@
...
+    However, if the `db_library` field, a single row of the PacBioRun table
+    class, is set via the constructor, the rest of the fields are populated
+    using this database row object, while  any other information passed to the
+    constructor is disregarded.
+
+    The  `db_library` field is not present in the model instance that is
+    returned by the constructor.
+    """
+
+    db_library: PacBioRun = Field(init_var=True)
```
Comment (nerdstrike) [inline lang_qc/models/pacbio/experiment.py:45]: Spot check - a "library" is the prep of one sample, therefore 1:1 with a row in `pac_bio_run`? As distinct from pool, which is a pool of libraries?

File: lang_qc/models/pacbio/well.py [hunk 4]
```diff
@@ -132,9 +133,10 @@ def pre_root(cls, values: dict[str, Any]) -> dict[str, Any]:
         """
 
         # https://github.com/pydantic/pydantic-core/blob/main/python/pydantic_core/_pydantic_core.pyi
-        mlwh_db_row: PacBioRunWellMetrics = values.kwargs["db_well"]
-        assert mlwh_db_row
+        if "db_well" not in values.kwargs:
```
Comment (nerdstrike) [inline lang_qc/models/pacbio/well.py:136]: The type hint says values is a dict. Does this work with just `if "db_well" not in values:`?

Comment (nerdstrike) [reply lang_qc/models/pacbio/well.py:136]: Also it seems strange that Pydantic can't just make `db_well` a required attribute without you having to check it.

File: tests/test_mlwh_db_classes.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        id2: f"pac_bio_run_name=TRACTION-RUN-1140, well_label=A1, plate_number=2, id_pac_bio_product={id2}",
+    }
+
+    for id in data.keys():
+        query = select(PacBioRunWellMetrics).where(
+            PacBioRunWellMetrics.id_pac_bio_product == id
+        )
+        db_row = mlwhdb_test_session.execute(query).scalar_one()
+        assert (
+            db_row.__repr__()
```
Comment (nerdstrike) [inline tests/test_mlwh_db_classes.py:22]: maybe `str(db_row)` rather than calling internal function?

**Review by nerdstrike — COMMENTED** [review]
Seems fine, a few questions to ease my concerns.

File: lang_qc/models/pacbio/experiment.py (hunk 2 above)
Comment (mgcam) [reply lang_qc/models/pacbio/experiment.py:31]: True. If we need another one like this for a different platform, the fields can go to lang_qc::models::library and the build method stays in this class or something like this. I had it like this, then decided against it for now

File: lang_qc/models/pacbio/experiment.py (hunk 3 above)
Comment (mgcam) [reply lang_qc/models/pacbio/experiment.py:45]: Correct. A pool might contain one library as long as the library is indexed. The term `pool` is only used with indexed libraries.

File: tests/test_mlwh_db_classes.py (hunk 5 above)
Comment (mgcam) [reply tests/test_mlwh_db_classes.py]: done

File: lang_qc/db/mlwh_schema.py (hunk 1 above)
Comment (mgcam) [reply lang_qc/db/mlwh_schema.py]: done

File: lang_qc/models/pacbio/well.py (hunk 4 above)
Comment (mgcam) [reply lang_qc/models/pacbio/well.py:136]: `values` seems to be of type [ArgsKwarg ](https://github.com/pydantic/pydantic-core/blob/66331bb1a4977d3d927fd26c6929dd6532d586ef/python/pydantic_core/_pydantic_core.pyi#L864), which does not seem to be a part of public API.

**Review by nerdstrike — APPROVED** [review]

**Pull request merged**